### DISCUSSION TITLE
Is there some example can illustrate difference between property assignment and setimmediatevalue?

### DISCUSSION THREAD

**Discussion by xwzheng1020** [cocotb/cocotb#3144] [Q&A]
In my test, I found sig.value = new_value and sig.setimmediatevalue(new_value) will get same result. But the document says: 
> The assignment syntax sig.value = new_value has the same semantics as [HDL](https://docs.cocotb.org/en/stable/glossary.html#term-HDL): writes are not applied immediately, but delayed until the next write cycle.

piece of my code: 
```
@cocotb.test()
async def my_first_test(dut):
    """Try accessing the design."""
    await cocotb.start(Clock(dut.clk, 10, units="ns").start())
    dut.data_in.value = 1
    await Timer(28, units="ns")
    # dut.data_in.value = 2
    dut.data_in.setimmediatevalue(2)
    await Timer(10, units="ns")
```

When simulating in modelsim with GUI=1, I found that `data_in` will change to 2 at 28ns in both situation.

**Comment by marlonjames** [comment]
https://github.com/cocotb/cocotb/wiki/Timing-Model#setimmediatevalue has some info.

**Comment by xwzheng1020** [comment]
According to 
> Normal writes in cocotb of the form dut.handle.value = 1 are cached until the start of the [value change](https://github.com/cocotb/cocotb/wiki/Timing-Model#value-change) state, such that if you read the signal value immediately following a normal value set, you would read the previous value.Using `setimmediatevalue` to set a signal's value changes the value immediately, such that if you read the signal value immediately following a call to setimmediatevalue, you would read the value you just wrote. 

the following code will get old value:
```
@cocotb.test()
async def my_first_test(dut):  
    cocotb.start_soon(Clock(dut.clk, 10, units="ns").start())
    dut.data_in.value = 1
    await Timer(30, units="ns")
    dut.data_in.value = 2
    dut._log.info(f'data_in: {dut.data_in.value}')  # output: data_in: 1
```

and 
```
@cocotb.test()
async def my_first_test(dut):
    cocotb.start_soon(Clock(dut.clk, 10, units="ns").start())
    dut.data_in.value = 1
    await Timer(30, units="ns")
    dut.data_in.setimmediatevalue(2)
    dut._log.info(f'data_in: {dut.data_in.value}')  # expected output: data_in: 2, but actually get 1
```

what went wrong?

**Reply by ktbarrett** [reply]
Your simulator must not support that behavior.
